lightning/src/routing/router.rs

```diff
@@ -149,9 +162,78 @@ impl cmp::PartialOrd for RouteGraphNode {
 struct DummyDirectionalChannelInfo {
 	cltv_expiry_delta: u32,
 	htlc_minimum_msat: u64,
+	htlc_maximum_msat: Option<u64>,
 	fees: RoutingFees,
 }
 
+// It's useful to keep track of the hops associated with the fees required to use them,
+// so that we can choose cheaper paths (as per Dijkstra's algorithm).
+#[derive(Clone)]
+struct PaymentHop {
+	route_hop: RouteHop,
+	/// Amount available for routing considering the following limitations:
+	/// - UTXO capacity
+	/// - htlc_maximum_msat (direction-dependent)
+	/// - use of the same channel by other paths per MPP (including fees paid there)
+	/// It does not take into account fees to be paid on the current hop, so
+	/// this amount *should* cover transferring a value AND paying fees (on this channel).
+	available_msat: u64,
+	/// Minimal fees required to route to the previous hop via any of its inbound channels.
+	src_lowest_inbound_fees: RoutingFees,
+	/// Fees of the channel used in this hop.
+	channel_fees: RoutingFees,
+	/// Total fee (useful to choose hops as we traverse the graph dest to source of the payment) consists of:
+	/// - fee paid for the use of the current channel (see channel_fees)
+	/// - fee required to reach the source node of the current channel (estimate, see src_lowest_inbound_fees)
+	/// - all the fees paid *after* this channel on the way to the destination
+	/// This value should be updated only in the context of the whole path, see update_amount_and_recompute_fees.
+	total_fee_msat: u64,
+}
+
+// Instantiated with a list of hops with correct data in them collected during path finding,
+// an instance of this struct should be further modified only via given methods.
+#[derive(Clone)]
+struct PaymentPath {
+	hops: Vec<PaymentHop>,
+}
+
+impl PaymentPath {
+	fn get_amount_transferred_msat(&self) -> u64 {
+		return self.hops.last().unwrap().route_hop.fee_msat;
+	}
+
+	fn get_total_fee_paid_msat(&self) -> u64 {
+		return self.hops.first().unwrap().total_fee_msat;
+	}
+
+	// If an amount transferred by the path is updated, the fees should be adjusted. 
+	// Any other way to change fees may result in an inconsistency.
+	fn update_amount_and_recompute_fees(&mut self, amount_msat: u64) {
+		let mut total_fee_paid_msat = 0 as u64;
+		for i in (1..self.hops.len()).rev() {
+			let cur_hop_amount_msat = total_fee_paid_msat + amount_msat;
+			let mut cur_hop = self.hops.get_mut(i).unwrap();
+			let proportional_fee_millions = cur_hop_amount_msat.checked_mul(cur_hop.channel_fees.proportional_millionths as u64);
+			if let Some(new_fee) = proportional_fee_millions.and_then(|part| {
+					(cur_hop.channel_fees.base_msat as u64).checked_add(part / 1000000) }) {
+	
```

Comment:
nit: here and the blank line below you have a stray tab - `git show` in most terminals should highlight this for you automatically.